src/_P088_Ping.ino

```diff
@@ -0,0 +1,258 @@
+    Device[deviceCount].PullUpOption = false;
+    Device[deviceCount].InverseLogicOption = false;
+    Device[deviceCount].DecimalsOnly = true;
+    Device[deviceCount].FormulaOption = false;
+    Device[deviceCount].SendDataOption = true;
+    Device[deviceCount].TimerOption = true;
+    break;
+  }
+
+  case PLUGIN_GET_DEVICENAME:
+  {
+    //return the device name
+    string = F(PLUGIN_NAME_088);
+    break;
+  }
+  case PLUGIN_GET_DEVICEVALUENAMES:
+  {
+    strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[0], PSTR(PLUGIN_VALUENAME1_088));
+    break;
+  }
+
+  case PLUGIN_WEBFORM_LOAD:
+  {
+    char hostname[PLUGIN_088_HOSTNAME_SIZE];
+    LoadCustomTaskSettings(event->TaskIndex, (byte*)&hostname, PLUGIN_088_HOSTNAME_SIZE);
+    addFormTextBox(String(F("Hostname")), F("p088_ping_host"), hostname, PLUGIN_088_HOSTNAME_SIZE - 2);
+    success = true;
+    break;
+  }
+
+  case PLUGIN_WEBFORM_SAVE:
+  {
+    char hostname[PLUGIN_088_HOSTNAME_SIZE];
+    strncpy(hostname,  WebServer.arg(F("p088_ping_host")).c_str() , sizeof(hostname));
+    SaveCustomTaskSettings(event->TaskIndex, (byte*)&hostname, PLUGIN_088_HOSTNAME_SIZE);
+    success = true;
+    break;
+  }
+
+  case PLUGIN_INIT:
+  {
+    if (!P088_data) {
+      P088_data = new P088_data_struct();
+      P088_data->m_IcmpPCB = raw_new(IP_PROTO_ICMP);
+      raw_recv(P088_data->m_IcmpPCB, PingReceiver, NULL);
+      raw_bind(P088_data->m_IcmpPCB, IP_ADDR_ANY);
+      for (int i = 0; i < PLUGIN_088_MAX_INSTANCES; i++) {
+        memset(&P088_data->destIPAddress[i], 0x0, sizeof(P088_data->destIPAddress[0]));
+        memset(&P088_data->idseq[i], 0x0, sizeof(P088_data->idseq[0]));
+      }
+      P088_data->instances = 1;
+    } else {
+      P088_data->instances++;
+    }
+    UserVar[event->BaseVarIndex] = 0;
+    success = true;
+    break;
+  }
+
+  case PLUGIN_EXIT:
+  {
+    P088_data->instances--;
+    if (P088_data->instances == 0) {
+      raw_remove(P088_data->m_IcmpPCB);
+      delete P088_data;
+    }
+    break;
+  }
+
+  case PLUGIN_READ:
+  {
+    char hostname[PLUGIN_088_HOSTNAME_SIZE];
+    IPAddress ip;
+
+    // Does we have unanswered pings?
+    if (P088_data->destIPAddress[event->TaskIndex].addr != 0) {
+      UserVar[event->BaseVarIndex]++;
+    }
+
+    success = true;
+    if (!WiFiConnected()) {
+      break;
+    }
+    LoadCustomTaskSettings(event->TaskIndex, (byte*)&hostname, PLUGIN_088_HOSTNAME_SIZE);
+
+    if (WiFi.hostByName(hostname, ip) == false) {
+      break;
+    }
+
+    P088_data->destIPAddress[event->TaskIndex].addr = ip;
+
+    /* Generate random ID & seq */
+    P088_data->idseq[event->TaskIndex] = random(UINT32_MAX);
+    u16_t ping_len = 32 + sizeof(struct icmp_echo_hdr);
+    struct pbuf *packetBuffer = pbuf_alloc(PBUF_IP, ping_len, PBUF_RAM);
+    if (packetBuffer == NULL) {
+      break;
+    }
+
+    struct icmp_echo_hdr * echoRequestHeader = (struct icmp_echo_hdr *)packetBuffer->payload;
+    ICMPH_TYPE_SET(echoRequestHeader, ICMP_ECHO);
```

Comment:
You don't check here if the `echoRequestHeader` is a nullptr.
In the `PingReceiver` callback you do check whether the cast was successful.
Is it not needed here?

Reply:
I added this checks as seen in some code and after looking to lwip code - seems no need for this checks even in PingReceiver. I can't see way how pbuf->payload might appear as null after i run pbuf_header. Or it works, or it will return non-null on failure of pbuf_header.
Also seems it is better to check packet length, for truncated packet in PingReceiver. Will fix that.

Reply:
With a C-style cast I also don't know how it should return anything other than the object, unless the void pointer is a nullptr.
When you have a pointer to a base class and try to downcast to a derived class, then it is a valid way to check if some object is of the type of the derived class you try to cast to.
But as far as I know, this doesn't work on casts of a void pointer.  (please correct me if I'm wrong, since I really would like to know if I'm correct here)